src/ray/core_worker/task_interface.h

```diff
@@ -43,45 +42,136 @@ struct ActorCreationOptions {
+        actor_language_(actor_language),
+        actor_definition_descriptor_(actor_definition_descriptor),
         actor_cursor_(ObjectID::FromBinary(actor_id.Binary())),
-        task_counter_(0) {}
+        task_counter_(0),
+        num_forks_(0) {}
+
+  ActorHandle(const ActorID &actor_id, const ActorHandleID &actor_handle_id,
+              const WorkerLanguage actor_language,
+              const ActorDefinitionDescriptor &actor_definition_descriptor,
+              const ObjectID &actor_cursor, int task_counter, int num_forks)
+      : actor_id_(actor_id),
+        actor_handle_id_(actor_handle_id),
+        actor_language_(actor_language),
+        actor_definition_descriptor_(actor_definition_descriptor),
+        actor_cursor_(actor_cursor),
+        task_counter_(task_counter),
+        num_forks_(num_forks) {}
 
   /// ID of the actor.
   const ray::ActorID &ActorID() const { return actor_id_; };
 
   /// ID of this actor handle.
   const ray::ActorHandleID &ActorHandleID() const { return actor_handle_id_; };
 
- private:
-  /// Cursor of this actor.
+  /// Language of the actor.
+  const ray::WorkerLanguage ActorLanguage() const { return actor_language_; };
+  /// Descriptor of actor definition.
+  /// e.g. class info for Java actor. module and class info for Python actor.
+  const ray::ActorDefinitionDescriptor &ActorDefinitionDescriptor() const {
+    return actor_definition_descriptor_;
+  };
+
+  /// The unique id of the last return of the last task.
+  /// It's used as a dependency for the next task.
   const ObjectID &ActorCursor() const { return actor_cursor_; };
 
+  /// The number of tasks that have been invoked on this actor.
+  const int TaskCounter() const { return task_counter_; };
+
+  /// The number of times that this actor handle has been forked.
+  /// It's used to make sure ids of actor handles are unique.
+  const int NumForks() const { return num_forks_; };
+
+  std::unique_ptr<ActorHandle> Fork() {
+    std::unique_lock<std::mutex> guard(mutex_);
+    auto new_handle = std::unique_ptr<ActorHandle>(new ActorHandle(
+        actor_id_, ComputeNextActorHandleId(actor_handle_id_, ++num_forks_),
+        actor_language_, actor_definition_descriptor_));
+    new_handle->actor_cursor_ = actor_cursor_;
+    new_actor_handles_.push_back(new_handle->actor_handle_id_);
+    return new_handle;
+  }
+
+  void Serialize(std::string *output) {
+    std::unique_lock<std::mutex> guard(mutex_);
+
+    ray::rpc::ActorHandle temp;
+    temp.set_actor_id(actor_id_.Binary());
+    temp.set_actor_handle_id(actor_handle_id_.Binary());
+    temp.set_actor_language((int)actor_language_);
+    for (auto &item : actor_definition_descriptor_) {
+      temp.add_actor_definition_descriptor(item);
+    }
+    temp.set_actor_handle_id(actor_handle_id_.Binary());
+    temp.set_actor_cursor(actor_cursor_.Binary());
+    temp.set_task_counter(task_counter_);
+    temp.set_num_forks(num_forks_);
+    temp.SerializeToString(output);
```

Comment:
`temp.SerializeToString(output);` doesn't need to be locked.

Reply:
The new code wrapped `ray::rpc::ActorHandle` in `ray::ActorHandle`. The comment doesn't apply any more.